jetstream_pt/layers.py

```diff
@@ -59,15 +81,280 @@ def __init__(self, in_features, out_features, bias=None, device=None):
+      scaler = self.weight_scaler.unsqueeze(-1)
+      if not self.is_symmetric:
+        zero_point = self.zero_point.unsqueeze(-1) / scaler
+      else:
+        zero_point = None
+      w_dequantized = dequantize_tensor(
+          self.weight.to(torch.bfloat16), scaler, zero_point
+      )
+      return F.linear(inputs, w_dequantized)
+
+
+class WeightOnlyBlockwiseQuantizedLinear(torch.nn.Module):
+
+  def __init__(
+      self,
+      in_features,
+      out_features,
+      bias=False,
+      device=None,
+      is_symmetric=True,
+      use_dot_general=False,
+      block_size=128,
+      n_bit=8,
+  ):
+    super().__init__()
+    self.in_features = in_features
+    self.out_features = out_features
+
+    # Use dot general instead of einsum
+    # Use dot general is slow now.
+    self.use_dot_general = use_dot_general
+    # Flatten einsum operands to 3D. XLA was slow if operands are 4D. But it's fixed now.
+    # Same perf as non flattened one now.
+    self.flatten = False
+
+    self.block_size = block_size
+    n_blocks = in_features // block_size
+
+    if self.use_dot_general:
+      weight = torch.ones(
+          (n_blocks, out_features, block_size), dtype=torch.int8, device=device
+      )
+    else:
+      weight = torch.ones(
+          (n_blocks, block_size, out_features), dtype=torch.int8, device=device
+      )
+    self.register_buffer("weight", weight)
+
+    weight_scaler = torch.ones(
+        (n_blocks, out_features), dtype=torch.bfloat16, device=device
+    )
+    self.register_buffer("weight_scaler", weight_scaler)
+
+    self.is_symmetric = is_symmetric
+    if not self.is_symmetric:
+      zero_point = torch.ones(
+          (n_blocks, out_features), dtype=torch.bfloat16, device=device
+      )
+      self.register_buffer("zero_point", zero_point)
+    else:
+      self.register_buffer("zero_point", None)
+
+    self.n_bit = n_bit
+    # Flag to enable dequantize weight first, then do matmul. Useful for debugging.
+    self.run_fake_quantize = False
+
+  def _load_quantized_weights(self, w_q, scale, zp=None):
+    """
+    Load weights quantized by 'quantize_tensor'.'
+    """
+    self.weight, self.weight_scaler, self.zero_point = load_q_weight_helper(
+        w_q, scale, zp, self.block_size
+    )
+
+  def quantize_weight_from_nn_linear(self, weight):
+    assert weight.dim() == 2, "Expect 2D weight from torch.nn.Linear."
+    assert weight.shape == (
+        self.out_features,
+        self.in_features,
+    ), f"Unexpected weight shape ({self.out_features}, {self.in_features})."
+    w_q, scale, zp = quantize_tensor(
+        weight, (1,), self.n_bit, self.is_symmetric, self.block_size
+    )
+    w_dq = dequantize_tensor(w_q, scale, zp)
+    print("check qweight cosine dist: ", _calc_cosine_dist(weight, w_dq))
+    # breakpoint()
+    self._load_quantized_weights(w_q, scale, zp)
+
+  @staticmethod
+  def blockwise_jax_kernel(j_inputs, j_weight, j_weight_scaler, j_zero_point):
```

Comment:
Is there any reason to have "j" as prefix of parameter name?

Reply:
Yes, this is passed into `torchjax.call_jax`, the parameters are jax tensors, so I name them with a prefix j.

Reply:
Thanks for sharing it! I feel it would be more readable if removing the prefix j and adding a comment to mention all these are jax tensor under this function.

Reply:
Updated